test/functional/feature_taproot.py

```diff
@@ -0,0 +1,638 @@
+            # Annex is always required for leaf version 0x50
+            # Unless the original version is 0x50, we couldn't convert it to 0x50 without using annex
+            tmp = damage_bytes(ret[1])
+            while annex is None and tmp[0] == ANNEX_TAG and ret[1][0] != ANNEX_TAG:
+                tmp = damage_bytes(ret[1])
+            ret[1] = tmp
+        else:
+            ret[0] = damage_bytes(ret[0])
+    if annex is not None:
+        ret += [annex]
+    # Randomly add input witness
+    if random.choice([True, False]):
+        for i in range(random.randint(1, 10)):
+            ret = [random_bytes(random.randint(0, MAX_SCRIPT_ELEMENT_SIZE * 2))] + ret
+    tx.wit.vtxinwit[input_index].scriptWitness.stack = ret
+
+def spender_sighash_mutation(spenders, info, comment, standard=True, **kwargs):
+    spk = info[0]
+    addr = get_taproot_bech32(info)
+
+    def fn(t, i, u, v):
+        return spend_single_sig(t, i, u, damage=not v, info=info, **kwargs)
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=standard, sat_function=fn))
+
+def spender_two_paths(spenders, info, comment, standard, success, failure):
+    spk = info[0]
+    addr = get_taproot_bech32(info)
+
+    def fn(t, i, u, v):
+        return spend_single_sig(t, i, u, damage=False, info=info, **(success if v else failure))
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=standard, sat_function=fn))
+
+def spender_alwaysvalid(spenders, info, comment, **kwargs):
+    spk = info[0]
+    addr = get_taproot_bech32(info)
+
+    def fn(t, i, u, v):
+        return spend_alwaysvalid(t, i, damage=not v, info=info, **kwargs)
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=False, sat_function=fn))
+
+def spender_alwaysvalid_p2sh(spenders, info, comment, standard, script):
+    spk = get_p2sh_spk(info)
+    addr = get_version1_p2sh(info)
+
+    def fn(t, i, u, v):
+        if v:
+            t.vin[i].scriptSig = CScript([info[0]])
+            # Empty control block is only invalid if we apply taproot rules,
+            # which we shouldn't if the spend is wrapped in P2SH
+            t.wit.vtxinwit[i].scriptWitness.stack = [script, bytes()]
+        else:
+            t.vin[i].scriptSig = CScript()
+        return
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=standard, sat_function=fn))
+
+def nested_script(script, depth):
+    if depth == 0:
+        return script
+    return [nested_script(script, depth - 1), CScript([OP_RETURN])]
+
+UTXOData = namedtuple('UTXOData', 'input,output,spender')
+
+class TaprootTest(BitcoinTestFramework):
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.extra_args = [["-whitelist=127.0.0.1", "-acceptnonstdtxn=0", "-par=1"]]
```

Comment:
FYI nonstandard txns are rejected by default in regtest now

Reply:
Done.